### PR TITLE
nixos/mastodon: add assertion for only allowing one sidekiq scheduler queue

### PR DESCRIPTION
**Pull request by erictapen**
###### Description of changes

This was meant to be included in https://github.com/NixOS/nixpkgs/pull/202408, but apparently I forgot to push it before merging.

This assertion makes sure, that there can be only one Sidekiq `scheduler` queue present, as upstream seems to be very cautious about it. See also this discussion: https://github.com/NixOS/nixpkgs/pull/202408#discussion_r1041082412

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/web-apps/mastodon.nix [hunk 1]
```diff
@@ -513,6 +513,17 @@ in {
...
         '';
       }
+      {
+        assertion = 1 == builtins.length
+          (lib.mapAttrsToList
+            (_: v: builtins.elem "scheduler" v.jobClasses || v.jobClasses == [ ])
+            cfg.sidekiqProcesses);
+        message = ''
+          There can be only one Sidekiq queue in
+          <option>services.mastodon.sidekiqProcesses</option> with jobClass
```
[the remark this answers is not in the document]
Comment (erictapen) [reply nixos/modules/services/web-apps/mastodon.nix]: Good to know! I removed the `<option>` tag.

File: nixos/modules/services/web-apps/mastodon.nix [hunk 2]
```diff
@@ -513,6 +513,17 @@ in {
...
+      {
+        assertion = 1 == builtins.length
+          (lib.mapAttrsToList
+            (_: v: builtins.elem "scheduler" v.jobClasses || v.jobClasses == [ ])
+            cfg.sidekiqProcesses);
+        message = ''
+          There can be only one Sidekiq queue in
+          services.mastodon.sidekiqProcesses with jobClass
+          "scheduler".
+        '';
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/mastodon.nix:525]:
```suggestion
        message = "There can be only one Sidekiq queue in services.mastodon.sidekiqProcesses with jobClass \"scheduler\".";
```

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by erictapen** [comment]
@SuperSandro2000 Is there a reason you approved, but didn't merge? Sometimes I get confused about the dos and dont's. Would you rather have a second person approve and merge this or would you like me as the maintainer to do the merge? I find myself self-merging urgent Mastodon updates sometimes, because the review capacity for Mastodon on NixOS is pretty low in general.

**Comment by SuperSandro2000** [comment]
Didn't have much of a reason

**Comment by jollheef** [comment]
Doesn't seem to work as intended. Test values are from the example.

```
# cat mastodon.nix | grep sidekiqProcesses -A17
    sidekiqProcesses = {
      all = {
        jobClasses = [ ];
        threads = null;
      };
      ingress = {
        jobClasses = [ "ingress" ];
        threads = 5;
      };
      default = {
        jobClasses = [ "default" ];
        threads = 10;
      };
      push-pull = {
        jobClasses = [ "push" "pull" ];
        threads = 5;
      };
    };
# nixos-rebuild switch
building Nix...
building the system configuration...
error:
       Failed assertions:
       - There must be one and only one Sidekiq queue in services.mastodon.sidekiqProcesses with jobClass "scheduler".
(use '--show-trace' to show detailed location information)
```

**Comment by erictapen** [comment]
@jollheef Ouch, thanks for letting me know. Care to review the fix? https://github.com/NixOS/nixpkgs/pull/227338

**Pull request merged**